pkg/integrations/prometheus/promql/parser.go

```diff
@@ -0,0 +1,564 @@
+		for _, entry := range lblEntry {
+			if entry.Name != "__name__" {
+				tagFilter := &structs.TagsFilter{
+					TagKey:          entry.Name,
+					RawTagValue:     entry.Value,
+					HashTagValue:    xxhash.Sum64String(entry.Value),
+					TagOperator:     segutils.TagOperator(entry.Type),
+					LogicalOperator: segutils.And,
+				}
+				mQuery.TagsFilters = append(mQuery.TagsFilters, tagFilter)
+			} else {
+				mQuery.MetricName = entry.Value
+			}
+		}
+	}
+
+	timeRange := &dtu.MetricsTimeRange{
+		StartEpochSec: startTime,
+		EndEpochSec:   endTime,
+	}
+
+	mQuery.OrgId = myid
+	mQuery.PqlQueryType = pqlQuerytype
+
+	intervalSeconds, err := mresults.CalculateInterval(endTime - startTime)
+	if err != nil {
+		return []*structs.MetricsQueryRequest{}, "", []*structs.QueryArithmetic{}, err
+	}
+
+	metricQueryRequest := &structs.MetricsQueryRequest{
+		MetricsQuery: mQuery,
+		TimeRange:    *timeRange,
+	}
+
+	mQueryReqs := make([]*structs.MetricsQueryRequest, 0)
+	mQueryReqs = append(mQueryReqs, metricQueryRequest)
+
+	queryArithmetic := make([]*structs.QueryArithmetic, 0)
+	var exitFromInspect bool
+
+	parser.Inspect(expr, func(node parser.Node, path []parser.Node) error {
+		if !exitFromInspect {
+			mQueryReqs, queryArithmetic, exitFromInspect, err = parsePromQLExprNode(node, mQueryReqs, queryArithmetic, intervalSeconds)
+		}
+		return err
+	})
+
+	if len(mQueryReqs) == 0 {
+		return mQueryReqs, pqlQuerytype, queryArithmetic, nil
+	}
+
+	mQuery = mQueryReqs[0].MetricsQuery
+
+	tags := mQuery.TagsFilters
+	for idx, tag := range tags {
+		var hashedTagVal uint64
+		switch v := tag.RawTagValue.(type) {
+		case string:
+			hashedTagVal = xxhash.Sum64String(v)
+		case int64:
+			hashedTagVal = uint64(v)
+		case float64:
+			hashedTagVal = uint64(v)
+		case uint64:
+			hashedTagVal = v
+		default:
+			log.Errorf("ParseMetricsRequest: invalid tag value type")
+		}
+		tags[idx].HashTagValue = hashedTagVal
+	}
+
+	if mQuery.MQueryAggs == nil {
+		mQuery.MQueryAggs = &structs.MetricQueryAgg{
+			AggBlockType:    structs.AggregatorBlock,
+			AggregatorBlock: &structs.Aggregation{AggregatorFunction: segutils.Avg},
+		}
+		mQueryReqs[0].MetricsQuery = mQuery
+	} else {
+		// If the first Block in the MQueryAggs is not an aggregator block, then add an default aggregator block with avg function
+
+		if mQuery.MQueryAggs.AggBlockType != structs.AggregatorBlock {
+			mQuery.MQueryAggs = &structs.MetricQueryAgg{
+				AggBlockType:    structs.AggregatorBlock,
+				AggregatorBlock: &structs.Aggregation{AggregatorFunction: segutils.Avg},
+				Next:            mQuery.MQueryAggs,
+			}
+			mQueryReqs[0].MetricsQuery = mQuery
+		}
+	}
+
+	return mQueryReqs, pqlQuerytype, queryArithmetic, nil
+}
+
+func parsePromQLExprNode(node parser.Node, mQueryReqs []*structs.MetricsQueryRequest, queryArithmetic []*structs.QueryArithmetic,
+	intervalSeconds uint32) ([]*structs.MetricsQueryRequest, []*structs.QueryArithmetic, bool, error) {
+	var err error = nil
+	exit := false
```

Comment:
What does `exit` mean? Why is it set to true after handling a `BinaryExpr`, sometimes true when handling `SubqueryExpr`, and false for handling other things?

Reply:
So the way `BinaryExpr` works... this is how that we are currently doing and I used the same here.

It can recursively calls the sub queries to be parsed inside a `BinaryExpr`. So it would mean, it would already be parsing all the components of Binary Expr `LHS` and `RHS`.

So there is no reason for us to go and parse the queries again.

The `pql.Inspect` will walk through each node in the `AST`. Which would mean it would go to first sub query in the Binary Expr, which we have already parsed in the `handleBinaryExpr` call. So I want to exit when the BinaryExpr call is returned.